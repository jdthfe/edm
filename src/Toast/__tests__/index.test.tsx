import { Toast } from '@src/index';
import React from 'react';
import { render, waitForElement, waitForElementToBeRemoved, fireEvent, wait, cleanup } from 'react-testing-library';

afterEach(cleanup);
describe('Toast', () => {
    const toastTxt = 'toastTxt';
    const duration = 1000;
    const time = 200;

    it('snapshots', async () => {
        let flag = false;
        const { baseElement } = render(<div />);
        const model = Toast.model({ children: 'normal', duration: 0, time });
        const alert = Toast.alert({ children: 'alert', duration: 0, time });
        const success = Toast.success({ children: 'success', duration: 0, time });
        const model2 = Toast.model();
        const alert2 = Toast.alert();
        const success2 = Toast.success();
        Toast.loading();
        setTimeout(() => {
            flag = true;
        }, 500);
        await wait(() => {
            if (!flag) throw flag;
        });

        //  排除 svg 文件干扰
        baseElement.querySelector('svg')!.remove();
        expect(baseElement).toMatchSnapshot('show');

        model();
        alert();
        success();
        model2();
        alert2();
        success2();
        Toast.closeLoading();

        flag = false;
        setTimeout(() => {
            flag = true;
        }, 500);
        await wait(() => {
            if (!flag) throw flag;
        });

        expect(baseElement).toMatchSnapshot('close');
    });
    const timeDeviation = 200;
    // 误差允许
    // it('Toast lasted for the specified time', async () => {
    //     const { getByText } = render(<div />);
    //     Toast.normal({ children: toastTxt, duration, time });

    //     const showTime = new Date().getTime();
    //     await waitForElement(() => getByText(toastTxt));
    //     await waitForElementToBeRemoved(() => getByText(toastTxt));
    //     const realDuration = new Date().getTime() - showTime;
    //     expect(realDuration >= duration && realDuration < duration + time + timeDeviation).toEqual(true);
    // });
    it('closed successfully', async () => {
        const { getByText } = render(<div />);
        const close = Toast.model({ children: toastTxt, duration });
        const showTime = new Date().getTime();
        await waitForElement(() => getByText(toastTxt));
        close();
        await waitForElementToBeRemoved(() => getByText(toastTxt));
        const realDuration = new Date().getTime() - showTime;
        expect(realDuration <= timeDeviation + time).toEqual(true);
    });

    const CoverClass = 'CoverClass';
    it('cover boolean', async () => {
        const { baseElement, getByText } = render(<div />);
        const mockFn = jest.fn();
        const close = Toast.model({
            children: toastTxt,
            coverProps: { onClick: mockFn, className: CoverClass },
        });
        await waitForElement(() => baseElement.querySelector('.' + CoverClass));
        fireEvent.click(baseElement.querySelector('.' + CoverClass)!);
        expect(mockFn).toBeCalled();
        close();
        await waitForElementToBeRemoved(() => baseElement.querySelector('.' + CoverClass));
        const close2 = Toast.model({
            children: toastTxt,
            noCover: true,
            coverProps: { onClick: mockFn, className: CoverClass },
        });
        await waitForElement(() => getByText(toastTxt));
        expect(baseElement.querySelector('.' + CoverClass)).toEqual(null);
        close2();
    });

    it('loading', async () => {
        const loadingText = 'loading';
        const { getByText, baseElement } = render(<div />);
        Toast.loading({ children: loadingText, duration, time, noCover: true, coverProps: { className: CoverClass } });

        const showTime = new Date().getTime();
        await waitForElement(() => getByText(loadingText));
        baseElement.querySelector('svg')!.remove();

        expect(baseElement.querySelector('.' + CoverClass)).toEqual(null);

        await waitForElementToBeRemoved(() => getByText(loadingText));
        const realDuration = new Date().getTime() - showTime;
        expect(realDuration >= duration && realDuration < duration + time + timeDeviation).toEqual(true);
    });

    it('base', async () => {
        const { baseElement } = render(<Toast visible>{toastTxt}</Toast>);
        expect(baseElement).toMatchSnapshot('show');
    });
});
