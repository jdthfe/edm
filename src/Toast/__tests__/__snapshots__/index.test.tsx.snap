// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast base: show 1`] = `
<body>
  
  
  

  
  
  

  <div />
  <div
    class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
  />
  <div
    class="eui-toast eui-fade-entry-done eui-fade"
  >
    toastTxt
  </div>
</body>
`;

exports[`Toast snapshots: close 1`] = `
<body>
  
  
  

  <div>
    <div />
  </div>
</body>
`;

exports[`Toast snapshots: show 1`] = `
<body>
  
  
  

  <div>
    <div />
  </div>
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      normal
    </div>
  </div>
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#alert"
        />
      </svg>
      <div
        class="eui-whitespace eui-whitespace-l"
      />
      alert
    </div>
  </div>
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#successful"
        />
      </svg>
      <div
        class="eui-whitespace eui-whitespace-l"
      />
      success
    </div>
  </div>
  <div />
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#alert"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#successful"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      class="eui-Cover eui-Cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl eui-icon-loading"
        fill="#fff"
      >
        <use
          xlink:href="#loading"
        />
      </svg>
    </div>
  </div>
</body>
`;
